keps/sig-windows/0000-20181221-windows-group-managed-service-accounts-for-container-identity.md

```diff
@@ -99,105 +101,152 @@ A developer has a Windows web app that they would like to deploy in a container
-9. During any subsequent pod update, any changes to a pod's `securityContext` is blocked (as is the case today) by `ValidatePodUpdate`.
+5. Application admins will deploy app pods that require a GMSA identity along with a Service Account authorized to use the GMSAs. There will be two options regarding how to specify the GMSACredSpec for pods and containers:
+
+  - Specify the name of the desired GMSACredSpec (e.g. `webapp1-credspec`): In the alpha stage of this feature, the name of the desired GMSACredSpec can be set through an annotation on the pod: `pod.alpha.kubernetes.io/windows-gmsa-credspec-name` as well as for each container through annotations of the form: `container.windows-gmsa-credspec-name.alpha.kubernetes.io/container-name`. In the beta stage, the annotations will be promoted to a field in the securityContext of the pod: `podspec.securityContext.windows.gmsaCredSpecName` and in the securityContext of each container:  `podspec.container[i].securityContext.windows.gmsaCredSpecName`. The GMSACredSpec name for a container will override the GMSACredSpec entries for the whole pod.
+
+  - Specify the contents of the `credpec` field of GMSACredSpec that gets passed down to the runtime: In the alpha stage of this feature, a JSON representation of the contents of the desired GMSACredSpec can be set through an annotation on the pod: `pod.alpha.kubernetes.io/windows-gmsa-credspec` as well as for each container through annotations of the form: `container.windows-gmsa-credspec.alpha.kubernetes.io/container-name`. In the beta stage, the annotations will be promoted to a field the securityContext of the pod `podspec.securityContext.windows.gmsaCredSpec` and in the securityContext of each  container: `podspec.container[i].securityContext.windows.gmsaCredSpec`.
+
+It is expected that users will typically use the first option as it is more user friendly. The ability to specify credspecs for each container within a pod aligns with how security attributes like `runAsGroup`, `runAsUser`, etc. can be specified at the pod level and overridden at the container level if necessary.
+
+6. A mutating webhook admission controller, GMSAAuthorizer, will act on pod creations and execute the following:
+
+  - Check the pod spec for references to names of GMSACredSpec objects in the pod (and per-container) annotations [for Alpha] or securityContext fields [Beta onwards]. Make sure the GMSACredSpec objects exist and use the contents of the `credspec` field of the GMSACredSpec to populate annotations `pod.alpha.kubernetes.io/windows-gmsa-credspec` and `container.windows-gmsa-credspec.alpha.kubernetes.io/container-name` [in Alpha] or `podspec.securityContext.windows.gmsaCredSpec` and `podspec.container[i].securityContext.windows.gmsaCredSpec` [Beta onwards]. If the specified names of GMSACredSpec cannot be resolved, pod creation will be failed with 404: Not Found.
```

Comment:
> If the specified names of GMSACredSpec cannot be resolved, pod creation will be failed with 404: Not Found.

The current behaviour in my PR is 400 bad request in that case. Happy to change if 404 seems more relevant, your call :)

Reply:
Not an expert but I was thinking 404 is slightly more specific (like 403 for authz failures).

Reply:
Fair enough, will change.